src/storage/LockingResourceStore.ts

```diff
@@ -58,81 +59,83 @@ export class LockingResourceStore implements AtomicResourceStore {
    * @param identifier - Identifier that should be locked.
    * @param func - Function to be executed.
    */
-  protected async lockedRun<T>(identifier: ResourceIdentifier, func: () => Promise<T>): Promise<T> {
-    const lock = await this.locks.acquire(identifier);
-    try {
-      return await func();
-    } finally {
-      await lock.release();
-    }
+  protected async lockedWriteRun<T>(identifier: ResourceIdentifier, func: () => Promise<T>): Promise<T> {
+    return this.locks.withWriteLock(identifier, func);
   }
 
   /**
-   * Acquires a lock for the identifier that should return a representation with Readable data and releases it when the
-   * Readable is read, closed or results in an error.
-   * When using this function, it is required to close the Readable stream when you are ready.
+   * Acquires a lock that is only released when all data of the resulting representation data has been read,
+   * an error occurs, or the timeout has been triggered.
+   * The resulting data stream will be adapted to renew the timer every time data is read.
+   *
+   * In case the data of the resulting stream is not needed it should be closed to prevent a timeout error.
    *
    * @param identifier - Identifier that should be locked.
    * @param func - Function to be executed.
    */
   protected async lockedRepresentationRun(identifier: ResourceIdentifier, func: () => Promise<Representation>):
   Promise<Representation> {
-    const lock = await this.locks.acquire(identifier);
-    let representation;
-    try {
+    // Create a new Promise that resolves to the resulting Representation
+    // while only unlocking when the data has been read (or there's a timeout).
+    return new Promise((resolve, reject): void => {
```

Comment:
This block is probably the one I spent the most time on so here's a full explanation why it is like this (and then you can tell me this way simpler solution I missed 😉 ).

The core problem is a combination of the following: 
 1. We want `getRepresentation` (and thus `lockedRepresentationRun`) to resolve with a Representation as output.
 2. We want the the lock to be released when all the data has been read.
 3. The read lock is released iff `withReadLock` resolves.
The combination of these factors means that `lockedRepresentationRun` needs to resolve before `withReadLock` resolves.

> Since `withReadLock` seems to return the return value of its callback.

Yes it does, but from the above it follows that its callback is not the result of `getRepresentation`, its entire callback is `getRepresentation` -> `createExpiringRepresentation` -> `waitForStreamToEnd`. But `lockedRepresentationRun` already needs to resolve after the second call since otherwise we would have a deadlock. This is why inside the callback passed to `withReadLock` the `resolve` call is done. This tells the function the result can be returned but does not unlock the lock yet. (This is the part that got a bit more complicated due to the new interface).

Reply:
Check; perhaps just briefly add a comment about that timing?